### ISSUE TITLE
Turbo mode boot

### ISSUE THREAD

**Issue by andreas-jonsson** [640-KB/GLaBIOS#54]
Perhaps there is something wrong on my side...? But I'm not able to boot in to turbo mode.

Should this note be reversed?
https://github.com/640-KB/GLaBIOS/blob/a927c22919a6806d9eb2eb6b08b72803b06937ed/src/GLABIOS.ASM#L1858
https://github.com/640-KB/GLaBIOS/blob/a927c22919a6806d9eb2eb6b08b72803b06937ed/src/GLABIOS.ASM#L1864

If `TURBO_TYPE` is `TURBO_STD` and `IF BOOT_SPEED EQ BOOT_TURBO` then we should `OR MASK PBTB` right?

Right now it seems like the PPI port B setup value for turbo is reversed even if type is set to `TURBO_STD`.

**Comment by 640-KB** [comment]
Yeah, the way I did that is probably not the most readable way...  below that if `TURBO_TYPE EQ TURBO_REV` it does an XOR on `MASK PBTB` bit to reverse it the PPI bit after you've picked `BOOT_TURBO` or `BOOT_NORMAL` so it adjusts to suit your board PPI if bit 2 `0` means turbo or bit 2 `1` means turbo, etc. The (arbitrary) way I defined standard vs. reverse is that for standard, PPI bit 2 is 0=turbo, 1=normal and reverse is 1=turbo, 2=normal.  

https://github.com/640-KB/GLaBIOS/blob/a927c22919a6806d9eb2eb6b08b72803b06937ed/src/GLABIOS.ASM#L1868-L1873

That said... you are saying this logic isn't working correctly if you have `TURBO_STD` and `BOOT_TURBO`?  

https://github.com/640-KB/GLaBIOS/blob/a927c22919a6806d9eb2eb6b08b72803b06937ed/src/GLABIOS.ASM#L573-L582

Oh, I think I see what you mean.  In this section, it will always set `BOOT_SPEED` to `BOOT_TURBO` if `TURBO_TYPE` isn't `TURBO_NONE` effectively overriding `BOOT_SPEED` _if_ you set that in your machine config. I suspect I intended that `BOOT_SPEED` wasn't meant to be set directly, rather inferred by `TURBO_NONE`, but yes that would prevent you from booting to NORMAL speed with either turbo type.  I suppose in theory, using the opposite `TURBO_TYPE` for your board would do that as well, but that's not quite intuitive.  Is that the issue you are seeing then?

640KB

**Comment by andreas-jonsson** [comment]
Hmm.. it could be just me being confused.. Sorry if that is the case :D

But if I explain it like this:
VirtualXT uses PPI port 61, bit 2 set, for indicating turbo enabled. That would be TURBO_STD = 1 if I'm not mistaken?
During powerup the "hardware" sets this bit order to run in turbo mode at startup.

But given this configuration https://github.com/640-KB/GLaBIOS/pull/55 it looks like the BIOS disables it during boot. Is this the intended behavior? Can I configure something in order to have turbo bit set at boot? (in my specific machine configuration)

Here is the location turbo is disabled for me:
```
;----------------------------------------------------------------------------;
; Set up PPI port B
;
 E0AC  B0 A8				MOV	AL, PPI_B_BOOT		; KB hold low+disable, NMI on, spkr off
 E0AE  E6 61				OUT	PPI_B, AL			; send to 8255 Port B
```

**Comment by andreas-jonsson** [comment]
> In this section, it will always set `BOOT_SPEED` to `BOOT_TURBO` if `TURBO_TYPE` isn't `TURBO_NONE`

isn't this doing the exact opposite? Note, it is not setting `PBTB`

https://github.com/640-KB/GLaBIOS/blob/a927c22919a6806d9eb2eb6b08b72803b06937ed/src/GLABIOS.ASM#L1854-L1858

**Comment by 640-KB** [comment]
Ahh okay, I see what you mean.  I think a big part of the confusion is due to my having these documented in the code inconsistently, and backwards to each other!

https://github.com/640-KB/GLaBIOS/blob/90a98c2f8494c60da8c0ef41029b7b5ff126459f/src/GLABIOS.ASM#L122-L123

vs.

https://github.com/640-KB/GLaBIOS/blob/90a98c2f8494c60da8c0ef41029b7b5ff126459f/src/GLABIOS.ASM#L1822

Side note... originally, the first turbo board that I came across used 0=turbo, 1=normal so I just named that "Standard".  However, after testing with more boards, it actually seemed like the opposite (that I named "Reverse") was more common with 1=turbo, 0=normal, so in retrospect it should have probably been named the other way around.  That is actually why `TURBO_TYPE = TURBO_REV` is the default in the code.

Anyway, my thinking here is, in order to minimize any changes to the code/behavior that might affect other build types, I'll review and fix all of the inline documentation to make sure it's consistent and correct.  I want to go back into old versions and see if it got changed at some point or if it's always been wrong.

> But given this configuration #55 it looks like the BIOS disables it during boot. Is this the intended behavior? Can I configure something in order to have turbo bit set at boot? (in my specific machine configuration)

To answer your actual question, in the VirtualXT configuration, remove the `TURBO_TYPE` override from the block, which will make it use the default (more common) `TURBO_REV` that actually means 1=turbo, 0=normal. See if that will do the trick?

Thx for bringing this to light!

640KB
